# [Phystech@DataScience](https://thetahat.ru/courses/ph-ds-2025-spr)
## Домашнее задание 8

**Правила, <font color="red">прочитайте внимательно</font>:**

* Выполненную работу нужно отправить телеграм-боту `@thetahat_pds_bot`. Для начала работы с ботом каждый раз отправляйте `/start`. Дождитесь подтверждения от бота, что он принял файл. Если подтверждения нет, то что-то не так. **Работы, присланные иным способом, не принимаются.**
* Дедлайн см. в боте. После дедлайна работы не принимаются кроме случаев наличия уважительной причины.
* Прислать нужно **ноутбук в формате `ipynb`**. Если вы строите интерактивные графики, их стоит прислать в формате html.
* Следите за размером файлов. **Бот не может принимать файлы весом более 20 Мб.** Если файл получается больше, заранее разделите его на несколько.
* Выполнять задание необходимо полностью самостоятельно. **При обнаружении списывания всем участникам списывания дается штраф -2 балла к итоговой оценке за семестр.**
* Решения, размещенные на каких-либо интернет-ресурсах, не принимаются. Кроме того, публикация решения в открытом доступе может быть приравнена к предоставлении возможности списать.
* Обратите внимание на <a href="https://thetahat.ru/courses/ai-rules"><b>правила использования ИИ-инструментов</b></a> при решении домашнего задания.
* **Код из рассказанных на занятиях ноутбуков** можно использовать без ограничений.
* Для выполнения задания используйте этот ноутбук в качестве основы, ничего не удаляя из него. Можно добавлять необходимое количество ячеек.
* Комментарии к решению пишите в markdown-ячейках.
* Выполнение задания (ход решения, выводы и пр.) должно быть осуществлено на русском языке.
* Решение проверяется системой ИИ-проверки <a href="https://thetahat.ru/"><img src="https://miptstats.github.io/theta_grader_small.png" style="display: inline; vertical-align: middle;"></a> **ThetaGrader**. Результат проверки валидируется и исправляется человеком, после чего комментарии отправляются студентам.
* Если код будет не понятен проверяющему, оценка может быть снижена.
* Никакой код из данного задания при проверке запускаться не будет. *Если код студента не выполнен, недописан и т.д., то он не оценивается.*

<b><font size="5" color="red">Важно!!! Правила заполнения ноутбука:</font></b>
* Запрещается удалять имеющиеся в ноутбуке ячейки, менять местами положения задач.
* Сохраняйте естественный линейный порядок повествования в ноутбуке сверху-вниз.
* Отвечайте на вопросы, а также добавляйте новые ячейки в предложенных местах, которые обозначены `<...>`.
* В markdown-ячейка, содержащих описание задачи, находятся специальные отметки, которые <font color="red"><b>запрещается модифицировать</b></font>.
* При нарушении данных правил работа может получить 0 баллов.

-----
**Баллы за задание:**

<b><font color="blue">Легкая часть</font></b> (достаточно на "хор"):
* Задача 1 &mdash; 20 баллов
* Задача 2 &mdash;  20 баллов
* Задача 3 &mdash;  25 баллов
* Задача 4 &mdash; 25 баллов

<b><font color="orange">Сложная часть</font></b> (необходимо на "отл"):
* Задача 5 &mdash; 35 баллов


-----

In [1]:
# Bot check

# HW_ID: phds_hw8
# Бот проверит этот ID и предупредит, если случайно сдать что-то не то.

# Status: not final
# Перед отправкой в финальном решении удали "not" в строчке выше.
# Так бот проверит, что ты отправляешь финальную версию, а не промежуточную.
# Никакие значения в этой ячейке не влияют на факт сдачи работы.

In [2]:
import numpy as np
import scipy.stats as sps
import matplotlib.pyplot as plt
import seaborn as sns

sns.set(palette='Set2')

### <b><i><font color="blue">Легкая часть</font></i></b> 


### Задача 1
Пусть $X = ( X_1, ... X_n)$ — выборка из неизвестного распределения $\mathsf{P} \in \{ \mathsf{P}_{\theta} | \theta \in \Theta \}$
и для проверки гипотез $\mathsf{H}_0 : \theta \in [−1, 1]$   vs.  $\mathsf{H}_1 : \theta \notin [−1, 1]$ используется критерий
$S = \{x \in \mathbb{R}^n | T(x) > 1\}$. Предположим, $T(x)$ имеет распределение $\mathcal{N}(\theta, 1)$. Какие
ответы возможны в результате проверки данных гипотез и в каких случаях? Найдите
вероятность ошибки первого рода.

*Указание:*
- Не забудьте, что в определении вероятности ошибки I рода стоит супремум по всем параметрам из основной гипотезы
- Можно показать, где достигается супремум графически. Например, с помощью `sps.norm`, где параметр `loc` отвечает за сдвиг.



Возможны следующие варианты в результате проверки данных гипотез:
1) $T(x) > 1$, тогда $x \in S$, значит, $H_0$ отвергается (статистически значимый результат).
2) $T(x) \leq 1$, тогда $x \notin S$, значит, $H_0$ $\textbf{не}$ отвергается (статистически незначимый результат).

Найдём вероятность ошибки первого рода, то есть вероятность отвергнуть $H_0$ при $\theta \in [-1,1]$. То есть нужно найти 
\begin{equation*}
\alpha = \sup_{\theta \in [-1, 1]} \mathbb{P}_\theta(T(x) > 1).
\end{equation*}

$T(x) ~ \mathcal{N}(\theta, 1)$, значит, $\mathbb{P}_\theta(T(x) > 1) = 1 - F(1-\theta)$, где $F$ - функция распределения для $\mathcal{N}(\theta, 1)$. Тогда 
\begin{equation*}
\alpha = \sup_{\theta \in [-1, 1]}  (1 - F(1-\theta)),
\end{equation*}
где супремум достигается при $\theta = 1$. Тогда
\begin{equation*}
\alpha = 1 - F(0) = 0.5.
\end{equation*}

### Задача 2

Проведен эксперимент, получены данные из экспоненциального распределения. 

In [3]:
sample = [0.11731702, 0.75253036, 0.32918642, 0.22823564, 0.04240622,
        0.04239907, 0.01495969, 0.50280772, 0.22977054, 0.30781252,
        0.00519983, 0.87588937, 0.44660739, 0.05967191, 0.05016975,
        0.05065286, 0.09068843, 0.18598196, 0.14138427, 0.08605575,
        0.23659272, 0.03755863, 0.08637888, 0.1140693 , 0.15223367,
        0.384484  , 0.05568397, 0.18050729, 0.22437618, 0.01189096]

Необходимо проверить, является ли это распределение с параметром $\lambda=2$. Используя Критерий Вальда, сделайте вывод по данному предположению.

$X_1, ... X_n$ &mdash; выборка из распределения $Exp(\lambda)$.  
Проверьте гипотезу $\mathsf{H}_0\colon \lambda = 2$ vs. $\mathsf{H}_1\colon \lambda \neq 2$  
  

 
 

**1.** Из лекции вы узнали про критерий Вальда.
Для случая двусторонней альтернативы $\mathsf{H}_1\colon \theta \neq \theta_0$ критерий имел следующий вид:
$$\large{S = \left\{ \left|\sqrt{n} \frac{\hat{\theta} - \theta_0}{\hat{\sigma}} \right| > z_{1 - \frac{\alpha}{2}} \right\}}$$

где $\hat{\theta}$ &mdash; асимптотически нормальная оценка $\theta$ с асимптотической дисперсией $\sigma^2(\theta)$, 
$\hat{\sigma}$ &mdash; состоятельная оценка $\sigma(\theta)$.  

Эквивалентный асимптотичсекий доверительный интервал для параметра $\theta$ уровня доверия $1-\alpha$
$$C = \left( \hat{\theta} - \frac{z_{1-\alpha/2} \hat{\sigma}}{\sqrt{n}}, \hat{\theta} + \frac{z_{1-\alpha/2} \hat{\sigma}}{\sqrt{n}}\right)$$

На первой лекции вы получали, что $\frac{1}{\overline{X}}$ &mdash; АНО для параметра $\theta$ c асимптотической дисперсией $\theta^2$


 **Выпишите состоятельную оценку дисперсии и статистику критерия Вальда**
 
 

**Ответ: Статистика критерия Вальда: $W(X) = \sqrt{n}\frac{\hat{\theta} - \theta_0}{\hat{\sigma}}$; 
Состоятельная оценка дисперсии: $\hat{\sigma}^2 = \hat{\theta}^2$** .

На лекции вы узнали про p-value &mdash; это вероятность получить при справедливости $H_0$ такое значение статистики $t = T(x)$ или еще более экстремальное, то есть в случае двустороннего критерия
$$p(x) = \mathsf{P}_0(T(X) \geq|t|) + \mathsf{P}_0(T(X) \leq -|t|)$$
Для расчета можно использовать функции из библиотеки `scipy.stats`.

**2**. Оформите функцию подсчета статистики критерия, p-value и доверительного интервала 

In [4]:
def wald_test(sample, theta, estimation_theta, estimation_sigma, alternative='two_sided'):
    """
    param sample: реализация выборки
    param theta: истинное значение параметра
    param estimation_theta: оценка параметра
    param estimation_sigma: оценка асимптотической дисперсии
        оценки estimation_sigma
    param alternative: вид альтернативной гипотезы,
        может принимать одно из значений 'two_sided', 'less', 'greater'

    return statistic
    return p_value
    return conf_int - доверительный интервал
    """
    alpha = 0.05
    statistic = np.sqrt(len(sample))*((estimation_theta - theta)/(estimation_sigma))
    if alternative == 'two_sided':
        z = sps.norm.ppf(1-alpha/2)
        conf_int = (estimation_theta - z*estimation_sigma/np.sqrt(len(sample)), estimation_theta + z*estimation_sigma/np.sqrt(len(sample)))
        p_value = sps.norm.cdf(-np.abs(statistic)) + sps.norm.sf(np.abs(statistic))
    elif alternative == 'greater':
        z = sps.norm.ppf(1-alpha)
        conf_int = (-np.inf, estimation_theta - z*estimation_sigma/np.sqrt(len(sample)))
        p_value = sps.norm.sf(statistic)
    elif alternative == 'less':
        z = sps.norm.ppf(alpha)
        conf_int = (estimation_theta + z*estimation_sigma/np.sqrt(len(sample)), np.inf)
        p_value = sps.norm.cdf(statistic)
    return statistic, p_value, conf_int
    

In [5]:
estimation_theta = 1/np.mean(sample)
theta = 2
estimation_sigma = estimation_theta

**3**. Проверьте гипотезу с двусторонней альтернативой.

In [6]:
two_sided_statistic, two_sided_p_value, two_sided_conf_int = wald_test(sample, theta, estimation_theta, estimation_sigma, 'two_sided')
print(f'Статистика Вальда: {two_sided_statistic}')
print(f'Доверительный интервал: {two_sided_conf_int}')
print(f'p_value: {two_sided_p_value}')

Статистика Вальда: 3.2704505437191247
Доверительный интервал: (3.187693843343649, 6.740324187101983)
p_value: 0.0010737631605120244


**Вывод:** Так как $2 \notin (3.19, 6.74)$, то есть не попадает в доверительный интервал, то можем отвергнуть гипотезу $H_0: \lambda = 2$. Однако что реализация статистики Вальда $3.19 < W(X) = 3.27 < 6.74$, значит мы не попадаем в критерий, и, кроме того p-value $\approx 0.001 << \alpha = 0.05$, то есть вероятность попасть в критерий, а значит отвергнуть гипотезу по критерию Вальда мала.

**4**. Проверьте гипотезу с правосторонней альтернативой.

In [7]:
greater_statistic, greater_p_value, greater_conf_int = wald_test(sample, theta, estimation_theta, estimation_sigma, 'greater')
print(f'Статистика Вальда: {greater_statistic}')
print(f'Доверительный интервал: {greater_conf_int}')
print(f'p_value: {greater_p_value}')

Статистика Вальда: 3.2704505437191247
Доверительный интервал: (-inf, 3.4732783303117714)
p_value: 0.0005368815802560122


**Вывод:** Здесь 2 $\in (-\infty, 3.47)$, значит, $H_0$ не отвергается, тот же вывод делается из того, что $W(X) = 3.27 < 3.47$ и сравнения полученного p-value $\approx 0.0005$ (то есть вероятности получить данное значение статистики Вальда или более экстремальное и попасть в критерий, а значит не попасть в интервал) и $\alpha = 0.05$.

**5**. Проверьте гипотезу с левосторонней альтернативой.

In [8]:
less_statistic, less_p_value, less_conf_int = wald_test(sample, theta, estimation_theta, estimation_sigma, 'less')
print(f'Статистика Вальда: {less_statistic}')
print(f'Доверительный интервал: {less_conf_int}')
print(f'p_value: {less_p_value}')

Статистика Вальда: 3.2704505437191247
Доверительный интервал: (3.4732783303117705, inf)
p_value: 0.999463118419744


**Вывод:** В данном случае 2 $\notin (3.47, +\infty)$, значит гипотеза $H_0: \lambda = 2$ отвергается. Кроме того, реализация статистики Вальда $W(X) = 3.27 \notin  (3.47, +\infty)$, то есть не попадает в доверительный интервал, значит лежит в критерии, и кроме того,  p-value $\approx 0.99 >> \alpha = 0.05$, то есть вероятность получить такое значение статистики Вальда или более экстремальное (значит, попасть в критерий) велика в сравнении с $\alpha$, то есть по критерию Вальда $H_0$ тоже отвергается.

### Задача 3

На практике часто рассматривают асимптотические критерии, например, критерий Вальда. Такие критерии контролируют вероятность ошибки I рода на уровне $\alpha$ только в пределе, однако для небольших выборок она может сильно отличаться от $\alpha$, причем нередко в большую сторону. Поэтому для таких критериев возникает необходимость в вычислении значения вероятности ошибки I рода или *реального уровня значимости* при конечной выборке.

Зачастую посчитать точное значение реального уровня значимости довольно сложно, поэтому вместо него вычисляют оценку с помощью сэмплирования по методу Монте-Карло.

Пусть гипотеза $\mathsf{H}_0$ простая, то есть $\mathsf{H}_0\colon \mathsf{P} = \mathsf{P}_0$. Для ее проверки по конечной выборке $X_1, ..., X_n$ используется критерий $S$. **Опишите схему оценки реального уровня** значимости критерия $S$ методом Монте-Карло.


**Ответ:**
1. Создаем, например, $B$ выборок из заданного распределения $P_0$.
2. Для каждой выборки считаем статистику критерия $S$ и смотрим, какое количество выборок $K$ отвергнуты по критерию.
3. Оцениваем реальный уровень значимости как $\hat{\alpha} = \frac{K}{B}$.

Если гипотеза $\mathsf{H}_0$ сложная, то есть $\mathsf{H}_0\colon \mathsf{P} \in \mathscr{P}_0$, то можно повторить такую процедуру несколько раз для разных $\mathsf{P} \in \mathscr{P}_0$.

Рассмотрим альтернативную гипотезу $\mathsf{H}_1\colon \mathsf{P} \in \mathscr{P}_1$. **Опишите схему оценки мощности** критерия $S$ методом Монте-Карло.

**Ответ:** Аналогично оценке реального уровня значимости.

Пусть $X_1, ..., X_n$ &mdash; выборка из пуассоновского распределения с параметром $\theta$. Рассмотрим гипотезы $\mathsf{H}_0\colon \theta = 1\ \ vs.\ \ \mathsf{H}_1\colon \theta \not= 1$ и критерий Вальда для их проверки.

**1.** **Оцените реальный уровень значимости** критерия для $ n = 50$.

Оценка параметра $\theta$ для Пуассоновского распределения имеет вид: $\hat{\theta} = \overline{X}$. Для дисперсии: $\hat{\sigma} = \overline{X}$. 

In [9]:
theta = 1
n = 50
B = 10000
alpha = 0.05

z = sps.norm.ppf(1 - alpha/2)
indicator = 0

for i in range(B):
    sample = np.random.poisson(theta,n)
    estimation_theta = np.mean(sample)
    W = np.sqrt(n)*(estimation_theta - theta) / estimation_theta
    if np.abs(W) > z:
            indicator += 1

estimation_alpha = indicator/B

print(f"Оценка для реального уровня значимости: {estimation_alpha}")

Оценка для реального уровня значимости: 0.0686


**Вывод:** Видим, что $\hat{\alpha} = 0.07 > \alpha =0.05$, значит действительно, при малых размерах выборки контролировать вероятность ошибки первого рода с помощью критерия Вальда не удаётся.

**2.** Пусть $n=50$. **Сравните мощности критериев**, соответствующих двусторонней, левосторонней и правосторонней альтернативам при $\theta=0.2$ и при $\theta=2$.

In [10]:
def criterion_power(theta, alternative):
    indicator = 0
    for i in range(B):
        sample = np.random.poisson(theta, n)
        estimation_theta = np.mean(sample)
        W = np.sqrt(n)*(estimation_theta - theta) / estimation_theta
        if alternative == 'two_sided':
            z_two_sided = sps.norm.ppf(1 - alpha/2)
            if np.abs(W) > z_two_sided:
                indicator += 1
        elif alternative == 'greater':
            z_greater = sps.norm.ppf(1 - alpha)
            if W > z_greater:
                indicator += 1
        elif alternative == 'less':
            z_less = sps.norm.ppf(alpha)
            if W < z_less:
                indicator += 1
    return indicator / B

for theta in [0.2, 2]:
    print(f'Мощности для theta={theta}:')
    print(f'Для двусторонней альтернативы:{criterion_power(theta, alternative= "two_sided")}')
    print(f'Для правосторонней альтернативы:{criterion_power(theta, alternative="greater")}')
    print(f'Для левосторонней альтернативы:{criterion_power(theta, alternative="less")}')

Мощности для theta=0.2:


C:\Users\maksa\AppData\Local\Temp\ipykernel_22924\206586027.py:6: RuntimeWarning: divide by zero encountered in scalar divide
  W = np.sqrt(n)*(estimation_theta - theta) / estimation_theta


Для двусторонней альтернативы:0.3552
Для правосторонней альтернативы:0.134
Для левосторонней альтернативы:0.3351
Мощности для theta=2:
Для двусторонней альтернативы:0.0109
Для правосторонней альтернативы:0.002
Для левосторонней альтернативы:0.0325


**Вывод:** Аналогично уровню значимости.




### Задача 4

**Профиль физика**

Период колебания математического маятника можно расчитать по следующей формуле:
$$T = 2\pi \sqrt{\frac{l}{g}}$$
Вы выполнили лабораторную работу по вычислению ускорения свободного падения. Длина нити маятника 50 см. В своей работе вы измеряли время 10 полных колебаний (10 периодов) с помощью обычного секундомера и получили такие значение: 




In [11]:
T_sample = [14.47, 14.46, 14.73, 14.74, 16.14, 12.6, 14.17, 12.83, 14.1, 12.81, 12.98, 
            13.84, 13.96, 13.6, 14.13, 15.32, 13.98, 13.96, 12.86, 15.46, 13.2, 15.26, 
            14.86, 14.49, 14.8, 12.92, 14.65, 14.47, 15.88, 15.04, 14.84, 15.12, 12.96, 
            16.2, 11.99, 15.57, 13.55, 13.16, 14.76, 13.79, 12.58, 14.61, 14.1, 14.54, 
            14.72, 14.87, 13.37, 14.04, 13.09, 14.7]

Получите массив величин g, используя формулу выше.

In [12]:
l = 0.5
g_sample = [4*(np.pi)**2*l/(T/10)**2 for T in T_sample]

Предположим, такие величины имеют нормальное распределение $\mathcal{N}(a, \sigma^2)$. Используйте критерий Вальда для проверки гипотезы $\mathsf{H}_0: a = 9.81$ vs $\mathsf{H}_1: a \neq 9.81$


Оценки для нормального распределения: $\hat{a} = \overline{X}$, где $X$ - выборка; $\hat{\sigma}^2 = \frac{1}{n}\displaystyle\sum_{i=1}^{n}(X_i - \overline{X})^2$. 

In [13]:
a = 9.81
estimation_a = np.mean(g_sample)
list_for_dispersion = [(np.mean(g_sample) - g)**2 for g in g_sample]
estimation_sigma = np.sqrt(np.mean(list_for_dispersion))

statistic, p_value, conf_int = wald_test(g_sample, theta=a, estimation_theta=a, estimation_sigma=estimation_sigma, alternative='two_sided')
alpha = 0.05
z = sps.norm.ppf(1-alpha/2)
print(np.abs(statistic) > z)


False


**Вывод:** Видим, что реализация статистики не попала в критерий, значит, по критерию Вальда гипотеза $H_0$ не отвергается.

**Профиль биология**

Испытывается эффективность препарата, понижающего температуру. Каждому пациенту измеряют температуру до и после приема препарата. В результате разность температур для испытуемых получилась равной: 


In [14]:
a_sample = [1.19, 0.84, 1.3, 0.76, 0.99, 1.3, 0.97, 0.91, 0.97, 0.99, 1.18, 0.93, 0.84, 1.07, 0.95,
     1.01, 1.04, 0.84, 1.0, 1.19, 1.31, 0.97, 1.1, 0.86, 1.02, 0.95, 0.93, 0.84, 0.85, 0.78]

Предположим, такие величины имеют нормальное распределение $\mathcal{N}(a, \sigma^2)$. С помощью критерия Вальда проверьте гипотезу: $\mathsf{H}_0 \colon a=0$ vs. $\mathsf{H}_1 \colon a > 0$.

In [15]:
<...>

SyntaxError: invalid syntax (3000440494.py, line 1)

**Вывод:**

### <b><i><font color="orange">Сложная часть</font></i></b>


###  Задача 5
1. Пусть $X = ( X_1, ... X_n)$ — выборка из неизвестного распределения $\mathsf{P} \in \mathcal{N}(a, \sigma^2)$. Постройте критерий для проверки гипотез $\mathsf{H}_0 : a = 0 $   vs.  $\mathsf{H}_1 : a > 0$ уровня значимости $\alpha$.

Бонусный сложный пункт: получите выражение для p-value аналогично правостороннему критерию Вальда

*Указание:*
- Используйте статистику $T(X) = \sqrt{n - 1} \frac{\overline{X}}{s}$, Которая пр справедливости $\mathsf{H}_0$ (т.е. при $a = 0$) имеет распределение Стьюдента с $n-1$ степенями свободы $T_{n-1}$ (это будет позднее выведено на лекции) 
- Подумайте, какой вид должен иметь критерий? При каких значениях $T(X)$ гипотезу $\mathsf{H}_0$ разумно отвергать?
- Этот критерий реализован в `scipy.stats.ttest_1samp`. Использовать для данной задачи его можно, например, следующим образом: `sps.ttest_1samp(sample, popmean=0, alternative='greater')`. В результате применения функция вернет значение статистики критерия, p-value и количество степеней свободы. Вы можете проверить свой ответ с ее помощью.

2. Пусть  $X = ( X_1, ... X_n)$ &mdash; выборка из неизвестного распределения $\mathsf{P}$ с конечным вторым моментом. Постройте асимптотический критерий для проверки гипотез $\mathsf{H}_0 \colon \mathsf{E} X = 0 $   vs.  $\mathsf{H}_1 \colon \mathsf{E} X > 0$ уровня значимости $\alpha$. Покажите, что при $n \to \infty$ этот критерий совпадает с критерием из первого пункта.

**Решение:**